google/oauth2/_client.py

```diff
@@ -129,32 +176,51 @@ def _token_endpoint_request_no_throw(
             if hasattr(response.data, "decode")
             else response.data
         )
-
-        if response.status == http_client.OK:
+        response_data = ""
+        try:
             # response_body should be a JSON
             response_data = json.loads(response_body)
-            break
-        else:
-            # For a failed response, response_body could be a string
-            try:
-                response_data = json.loads(response_body)
-                error_desc = response_data.get("error_description") or ""
-                error_code = response_data.get("error") or ""
-                if (
-                    any(e == "internal_failure" for e in (error_code, error_desc))
-                    and retry < 1
-                ):
-                    retry += 1
-                    continue
-            except ValueError:
-                response_data = response_body
-            return False, response_data
-
-    return True, response_data
+        except ValueError:
+            response_data = response_body
+        return response, response_data
+
+    response, response_data = _perform_request()
+
+    if response.status == http_client.OK:
+        return True, response_data, None
+
+    retryable_error = _can_retry(
+        status_code=response.status, response_data=response_data
+    )
+
+    if not retryable_error or not should_retry:
```

Comment:
This condition makes me think whether we should rename `should_retry` to something like `can_retry` because even if user says `should_retry` `true`, we will not retry if error code says its not retryable.

Reply:
That sounds good to me, `should_retry` felt a bit off.